### PR TITLE
📖 Clarify template conventions and template reconciliation rules in the ClusterClass proposal

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR improves the ClusterClass proposal could be improved in explaining impact for the control plane providers, detailing 
- impacts on different control plane flavours:
- conventions for implementing new template types 
- expected behaviour for template <-> object reconciliation

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/5018

### CODE REVIEW

**Review by vincepri — APPROVED** [review]
LGTM

**Comment by vincepri** [comment]
Needs a rebase @fabriziopandini

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 1]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+      otherwise creating an object derived from a template will fail.
+
+**Note:** The existing InfrastructureMachineTemplate and BootstrapMachineTemplate objects already
+comply those conventions via explicit rules implemented in the code or via operational practices
+(otherwise creating machines will not be working already today).
+
+**Note:** As per this proposal, the definition of ClusterClass is immutable. The CC definition consists 
+of infrastructure object references, say AWSMachineTemplate, which could be immutable. For such immutable
+infrastructure objects, hard-coding the image identifiers leads to those templates being tied to a particular
+kubernetes version, thus making kubernetes version upgrades impossible. Hence, when using CC, infrastructure
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:519]:
```suggestion
Kubernetes version, thus making Kubernetes version upgrades impossible. Hence, when using CC, infrastructure
```

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 2]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+- For the fields existing only in the object but not in the corresponding template:
+    - Fields must be optional or a default value must be automatically assigned,
+      otherwise creating an object derived from a template will fail.
+
+**Note:** The existing InfrastructureMachineTemplate and BootstrapMachineTemplate objects already
+comply those conventions via explicit rules implemented in the code or via operational practices
+(otherwise creating machines will not be working already today).
+
+**Note:** As per this proposal, the definition of ClusterClass is immutable. The CC definition consists 
+of infrastructure object references, say AWSMachineTemplate, which could be immutable. For such immutable
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:517]: I wonder if we should mention that we treat the InfrastructureMachineTemplates as immutable even if they are mutable (i.e. we don't rollout changes to them)

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 3]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+infrastructure objects, hard-coding the image identifiers leads to those templates being tied to a particular
+kubernetes version, thus making kubernetes version upgrades impossible. Hence, when using CC, infrastructure
+objects MUST NOT have mandatory static fields whose values prohibit version upgrades.
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:520]: I wonder if we should phrase it a bit differently. 

I think the main problem is:

We don't want InfrastructureMachineTemplates to be tied to specific Kubernetes version, because otherwise we would have to update the ref to a InfrastructureMachineTemplate (to another one which supports the new Kubernetes version) in the ClusterClass **at the same time** when we update the Kubernetes version in all Clusters which are using the ClusterClass, which is impossible. Thus we need Kubernetes versions to be independent of the templates.

tl;dr compared to the current text. The immutability isn't the biggest problem, the main problem is that we would have to change the ref to a new template and the Kubernetes version both at the same time

I think this argumentation holds now and when we make the ClusterClass mutable. Of course both are correct.

Hope that all makes sense, feel free to keep it as is or modify it a bit :)

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 4]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+
+#### Notes on template <-> object reconciliation
+
+One of the key point of this proposal is that cluster topologies are continuously
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:524]:
```suggestion
One of the key points of this proposal is that cluster topologies are continuously
```

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 5]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+reconciled with the original templates to ensure consistency over time and to support changing the generated
+topology when necessary.
+
+Cluster Class and manged topologies reconciliation leverages on the conventions for template types documented in the previous
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:528]:
```suggestion
Cluster Class and managed topologies reconciliation leverages on the conventions for template types documented in the previous
```

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 6]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+paragraph and on the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+in order to reconcile objects in the cluster with the corresponding template in the ClusterClass.
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:530]:
```suggestion
in order to reconcile objects in the cluster with the corresponding templates in the ClusterClass.
```

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 7]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+Please note that:
+- If a field does not have a value in the template, but the users/another controller sets this value in the object,
+  the reconcile topology process will preserve the field value.
+- If a field does not exist in the template, but it exists only in the object, and the users/another controller
+  sets this field's value, the reconcile topology process will preserve it.
+- If a field is defined as `omitempty` but the field type is not a pointer or a type that has a built-in nil value, 
+  e.g a `Field1 bool 'json:"field1,omitempty"'`, it won't be possible to enforce the zero value in the template for the field, 
+  e.g `field1: false`, because the field is going to be removed and technically there is no way to distinguishing
+  unset from the zero value for that type.
+  NOTE: this is a combination not compliant with the recommendations in the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required)),
```
Comment (sbueringer) [inline docs/proposals/202105256-cluster-class-and-managed-topologies.md:551]:
```suggestion
  NOTE: this is a combination not compliant with the recommendations in the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required).
```

**Review by sbueringer — COMMENTED** [review]
Really nice work! :) No major findings from my side.

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 8]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
-
-The CAPI Cluster Controller would use this template to instantiate a control plane object for the topology. The current CRD contract mandates that the control plane provider supports `replicas` in the `spec` fields. With the introduction of ClusterClass which is responsible for handling the kubernetes version too, the current contract needs to be expanded to include the support for `version` in the `spec` fields.
+**Impact on the bootstrap providers**:
+- None.
+
+**Impact on the controlPlane providers**:
+- the provider implementers are required to implement the ControlPlaneTemplate type (e.g. `KubeadmControlPlaneTemplate` etc.).
+- it is also important to notice that:
+    - ClusterClass and managed topologies can work **only** with control plane providers implementing support for the `spec.version` field;
+      Additionally, it is required to provide support for the `status.version` field reporting the minimum
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply docs/proposals/202105256-cluster-class-and-managed-topologies.md:479]: https://github.com/kubernetes-sigs/cluster-api/issues/5025 😄

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md [hunk 9]
```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
...
+
+- Template fields must match or be a subset of the corresponding generated object.
+- A template can't accept values which are not valid for the corresponding generated object,
+  otherwise creating an object derived from a template will fail.
+  
+Objects generated from the template:
+
+- For the fields existing both in the object and in the corresponding template:
+    - The object can't have additional validation rules than the template,
+      otherwise creating an object derived from a template could fail.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply docs/proposals/202105256-cluster-class-and-managed-topologies.md:507]: Done; given how reconcile is going to be implemented (reconcile only what is set in the template), this is less relevant now.

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md (hunk 2 above)
Comment (fabriziopandini) [reply docs/proposals/202105256-cluster-class-and-managed-topologies.md:517]: This is a copy paste from the orginal proposal and in this PR I'm just reshuffling and but not changing this bits.
However, I think that what are you asking is implicit in "which could be mutatble". Also, the reconcile section below already clarifies how we are treating templates<->object reconciliation.

File: docs/proposals/202105256-cluster-class-and-managed-topologies.md (hunk 3 above)
Comment (fabriziopandini) [reply docs/proposals/202105256-cluster-class-and-managed-topologies.md:520]: This is a copy paste from the orginal proposal and in this PR I'm just reshuffling and but not changing this bits.
Given that this is going to be addressed when implementing mutability transformations/patches in the next iteration and the note will be removed soon, I would like to avoid to change a phrasing we all agreed upon.

**Review by sbueringer — COMMENTED** [review]
last nit, apart from that lgtm

**Review by vincepri — APPROVED** [review]

**Review by vincepri — APPROVED** [review]
/approve
